### PR TITLE
Feature/quiz module

### PR DESCRIPTION
**Pull request by zeala**
Hi,

I've started implementing the Quiz functionality. It's far from ready, but maybe you can have a quick review to see if this is on the right track. 

Thanks.

### CODE REVIEW

**Comment by brylie** [comment]
Just a quick look through the code brings up a few ideas. The quiz functionality might be implemented using Aldeed [AutoForm](https://github.com/aldeed/meteor-autoform) / [SimpleSchema](https://github.com/aldeed/meteor-simple-schema). 

The Quiz questions/question types would each be defined via a simple schema that would look something like this:

``` js
QuizSchema = new SimpleSchema({
    quizType: {
       type: String,
        allowedValues: ['multi-choice', 'freeform', ...]
        // Extend this list as new question types are added
    },
    multiChoice: {
        type: Object
        optional: true,
        customValidation: function () {
          // field is required only if quizType is set to 'multi-choice'
        }
    },
    "multiChoice.question" { // Question text to display to the user
        type: String
    }
    "multiChoice.choice": {
        type: [Object], // brackets here mean the question can have multiple choice
    },
    "multiChoice.choice.text": {
        type: String // Text to show user, i.e. answer text
    },
    "multiChoice.choice.isCorrectAnswer": {
        type: Boolean // It may be possible to have multiple correct answers
    }
});
```

This approach can help us to quickly have a form-based user interface for creating quizzes. The multi-choice approach can be made flexible to allow multiple correct answers, meaning the user has to select all correct answers for success. 

This schema can be extended to include other question types.

**Comment by Rahbaran** [comment]
i had a quick chat with @zeala . She's going to try to use autoform next week.

**Comment by brylie** [comment]
Great :-) I am available for pair-programming this week. @zeala, please let me know what day/time works, if you are interested in pair programming.

**Comment by Rahbaran** [comment]
@brylie : @zeala refactored her code to use autoform and schema. it would be great if you can have a look at it and to check if she's on the right track.

File: both/quizModel.js [hunk 1]
```diff
@@ -0,0 +1,197 @@
+Quiz = function(){
+    var quiz = this;
+    this.addNewQuestion = function(val){
+        //check if the question already exists
+        if (quiz.questions == undefined){
+            quiz.questions = [];
+        }
+        for (var q in quiz.questions){
+
```
Comment (brylie) [inline both/quizModel.js]: Cleanup this unused block.

File: both/quizModel.js [hunk 2]
```diff
@@ -0,0 +1,197 @@
+Quiz = function(){
+    var quiz = this;
+    this.addNewQuestion = function(val){
+        //check if the question already exists
+        if (quiz.questions == undefined){
```
Comment (brylie) [inline both/quizModel.js:5]: Use `===` comparison here.

File: both/quizModel.js [hunk 3]
```diff
@@ -0,0 +1,197 @@
...
+            quiz.questions = [];
+        }
+        for (var q in quiz.questions){
+
+        };
+
+        quiz.questions.push(val);
```
Comment (brylie) [inline both/quizModel.js:12]: Move this in to the `for` block above.

**Comment by brylie** [comment]
There are almost zero comments in the code. Please add a comment to _every_ significant line of code. This helps when people, including the original developer, review the code.

File: both/collections/questions.js [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        label: "Options",
+        type: [Object],
+        optional: false,
+        autoform: {
+            type: "radio-with-text-input",
+            options: function(){
+                var optionsArray = [];
+
+                for ( var i=0; i < 8; i++){
+                    var option = Quiz.generateAnswerOption("", false, i)
```
Comment (brylie) [inline both/collections/questions.js:42]: What does this line do. Please add comment in plain English.

File: both/collections/questions.js [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+    optionTitles: {
+        label: "Options",
+        type: [Object],
+        optional: false,
+        autoform: {
+            type: "radio-with-text-input",
+            options: function(){
+                var optionsArray = [];
+
+                for ( var i=0; i < 8; i++){
```
Comment (brylie) [inline both/collections/questions.js:41]: Why is this loop hard-coded to 8? You can use the `max` property of SinpleSchema arrays to set an upper limit.

File: both/collections/questions.js [hunk 6]
```diff
@@ -0,0 +1,91 @@
+Questions = new Mongo.Collection('questions');
+
+QuestionSchema = new SimpleSchema({
```
Comment (brylie) [inline both/collections/questions.js:3]: Attach this schema to the `Questions` object:

``` js
Questions.schema = new SimpleSchema({ ... });
```

File: both/collections/questions.js [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    },
+    title: {
+        type:String,
+        optional: false
+    },
+    description : {
+        type:String,
+        optional: false
+    },
+    numberOfOptions: {
```
Comment (brylie) [inline both/collections/questions.js:25]: Consider re-organizing your options schema as follows:

E.g. 

``` js
"options": {
  type: [Object],
  max: 8,
},
"options.$.title": {
  type: String
},
"options.$.isCorrectAnswer": {
  type: Boolean
}
```

File: both/collections/questions.js [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+        type: Number,
+        optional: true,
+        max: 10,
+        min: 2
+    },
+    optionTitles: {
```
Comment (brylie) [inline both/collections/questions.js:32]: You can remove this field, with the above refactoring.

File: both/collections/questions.js [hunk 9]
```diff
@@ -0,0 +1,91 @@
+Questions = new Mongo.Collection('questions');
+
+QuestionSchema = new SimpleSchema({
+    id: {
+        type: String,
+        optional: false,
+        unique: true
+    },
+    questionType: {
```
Comment (brylie) [inline both/collections/questions.js:9]: This field should probably have some `allowedValues` so that it is rendered in a select element. E.g. we only want to allow question types that we support, such as "multiple choice".

File: both/collections/questions.js [hunk 10]
```diff
@@ -0,0 +1,91 @@
+Questions = new Mongo.Collection('questions');
+
+QuestionSchema = new SimpleSchema({
+    id: {
```
Comment (brylie) [inline both/collections/questions.js:4]: This is unnecessary, as all MongoDB documents have an `_id` property.

File: both/collections/questions.js [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                    selectionFound = true;
+                }
+            }
+            if (!selectionFound){
+                return "invalidQuestionSelection";
+            }
+        }
+    },
+
+    "optionTitles.$.title": {
```
Comment (brylie) [inline both/collections/questions.js:69]: This field is unnecessary, if you refactor the "options" object.

E.g.

``` js
"options.$.title": {
  type: String
},
```

File: both/collections/questions.js [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+        type:String,
+        optional: false
+    },
+
+    "optionTitles.$.isSelected": {
```
Comment (brylie) [inline both/collections/questions.js:74]: This field is unnecessary if you simplify the `options` field. E.g.

``` js
"options.$.isCorrectAnswer": {
  type: Boolean
}
```

File: both/collections/questions.js [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+        type:Boolean,
+        optional: false
+    },
+    "optionTitles.$.index":{
```
Comment (brylie) [inline both/collections/questions.js:78]: This field may be unnecessary, since JavaScript arrays are natively indexed.

File: both/collections/questions.js [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+        type:Number,
+        optional: false
+    },
+
+
+    options:{
```
Comment (brylie) [inline both/collections/questions.js:84]: Please refactor this object as indicated:

``` js
"options": {
  type: [Object],
  max: 8,
},
"options.$.title": {
  type: String
},
"options.$.isCorrectAnswer": {
  type: Boolean
}
```

File: both/collections/quizzes.js [hunk 15]
```diff
@@ -0,0 +1,35 @@
+Quizzes = new Mongo.Collection('quizzes');
+
+QuizzesSchema.AnswerOptionSchema = new SimpleSchema({
```
Comment (brylie) [inline both/collections/quizzes.js:3]: This may be redundant, as individual options may be marked as `correctAnswer` directly.

File: both/collections/quizzes.js [hunk 16]
```diff
@@ -0,0 +1,35 @@
...
+        defaultValue: ""
+    },
+    isCorrect: {
+        type: Boolean,
+        defaultValue: false
+    },
+});
+
+//This schema will validate the initial creation of a quiz
+QuizzesSchema.QuizzesSchema = new SimpleSchema({
```
Comment (brylie) [inline both/collections/quizzes.js:16]: Define this schema directly in the Quizzes object:

``` js
Quizzes.schema = new SimpleSchema({ 
  ... 
});
```

File: both/common.js [hunk 17]
```diff
@@ -1,3 +1,6 @@
 SimpleSchema.messages({
-    "passwordMismatch": "Passwords do not match"
+    "passwordMismatch": "Passwords do not match",
+    "invalidQuestionSettings": "Invalid question settings",
```
Comment (brylie) [inline both/common.js:3]: We need to remember to internationalize these strings, so they can be translated.

File: both/quizModel.js [hunk 18]
```diff
@@ -0,0 +1,197 @@
+Quiz = function(){
+    var quiz = this;
+    this.addNewQuestion = function(val){
```
Comment (brylie) [inline both/quizModel.js:3]: This type of functionality is handled in a simple manner by embracing AutoForm.

File: both/quizModel.js [hunk 19]
```diff
@@ -0,0 +1,197 @@
...
+        for (var q in quiz.questions){
+
+        };
+
+        quiz.questions.push(val);
+    };
+
+};
+
+Quiz.convertToQuizObject = function(object){
```
Comment (brylie) [inline both/quizModel.js:17]: AutoForm and SimpleSchema handle this automatically.

File: both/quizModel.js [hunk 20]
```diff
@@ -0,0 +1,197 @@
...
+    quiz.questions = object.questions;
+    quiz.userAttempts = object.userAttempts;
+
+    return quiz;
+};
+
+Quiz.generateQuestion = function(questionType, quizId ){
+    var question = new Object();
+    question.quizId = quizId;
+    question.id = Random.id(); //assign an id to the question - need this for checking and validating answers
```
Comment (brylie) [inline both/quizModel.js:34]: MongoDB automatically assigns an `_id` property to each document.

File: both/quizModel.js [hunk 21]
```diff
@@ -0,0 +1,197 @@
...
+    quiz._id = object._id;
+    quiz.title = object.title;
+    quiz.lessonID = object.lessonID;
+    quiz.questions = object.questions;
+    quiz.userAttempts = object.userAttempts;
+
+    return quiz;
+};
+
+Quiz.generateQuestion = function(questionType, quizId ){
```
Comment (brylie) [inline both/quizModel.js:31]: AutoForm and SimpleSchema handle this automatically.

File: both/quizModel.js [hunk 22]
```diff
@@ -0,0 +1,197 @@
...
+    var question = new Object();
+    question.quizId = quizId;
+    question.id = Random.id(); //assign an id to the question - need this for checking and validating answers
+    question.questionType = questionType;
+
+    //if the question type is true-or-false, populate the answer options
+    if (question.questionType == QuizOptions.TRUE_OR_FALSE){
```
Comment (brylie) [inline both/quizModel.js:38]: We can use SimpleSchema to dynamically define what field to display, based on the `question.type` field.

File: both/quizModel.js [hunk 23]
```diff
@@ -0,0 +1,197 @@
...
+        question.optionTitles.push(falseOption);
+    };
+
+    question.description = "";
+    question.title = "";
+    question.options = [];
+    return question;
+};
+
+Quiz.generateAnswerOption = function (title, isSelected, index){
```
Comment (brylie) [inline both/quizModel.js:53]: SimpleSchema and AutoForm handle this simply and automatically.

File: both/quizModel.js [hunk 24]
```diff
@@ -0,0 +1,197 @@
...
+    var option = {};
+    option.title = title;
+    option.isSelected = isSelected;
+    option.index = index;
+
+    return option
+}
+
+
+Object.defineProperty(Quiz, "_title", {
```
Comment (brylie) [inline both/quizModel.js:63]: SimpleSchema handles this automatically. Then, we use plain JavaScript to access object properties.

File: both/quizModel.js [hunk 25]
```diff
@@ -0,0 +1,197 @@
...
+    get: function title(){
+        return this._title;
+    },
+    set: function title(val){
+        this._title = val;
+    }
+});
+Object.defineProperty(Quiz, "_questions", {
```
Comment (brylie) [inline both/quizModel.js:71]: SimpleSchema handles this automatically. Then, we use plain JavaScript to access object properties.

File: both/quizModel.js [hunk 26]
```diff
@@ -0,0 +1,197 @@
...
+    get: function question(){
+        return this._questions;
+    },
+    set: function questions(val){
+        this._questions = val;
+    }
+});
+
+Object.defineProperty(Quiz, "_lessonID", {
```
Comment (brylie) [inline both/quizModel.js:80]: SimpleSchema handles this automatically. Then, we use plain JavaScript to access object properties.

File: both/quizModel.js [hunk 27]
```diff
@@ -0,0 +1,197 @@
...
+    get: function lessonID(){
+        return this._lessonIDs
+    },
+    set: function lessonID(val){
+        this._lessonID = val;
+    }
+});
+
+Object.defineProperty(Quiz, "_userAttempts", { //an array of objects containig user ids, date and result
```
Comment (brylie) [inline both/quizModel.js:89]: We should move `userAttempts` into a separate collection. E.g.

``` js
const UserAttempts = new Mongo.collection("userAttempts");

UserAttempts.schema = new SimpleSchema({
  userId: {
    type: String,
    regEx: SimpleSchema.RegEx.Id
  },
  quizId: {
    type: String,
    regEx: SimpleSchema.RegEx.Id
  },
  userAttempts: {
    type: [Object] // Hold an array of all user attempts
  },
  "userAttempts.$.successful": {
    type: Boolean // true if user passes, false if user fails
  }
});
```

File: both/quizModel.js [hunk 28]
```diff
@@ -0,0 +1,197 @@
...
+        this._userAttempts = val;
+    }
+})
+
+
+
+
+QuizOptions = {};
+
+QuizOptions.MULTIPLE_CHOICE_SINGLE_ANSWER = "Multiple Choice - single answer";
```
Comment (brylie) [inline both/quizModel.js:103]: All of this can probably be done at the schema level.

File: both/quizModel.js [hunk 29]
```diff
@@ -0,0 +1,197 @@
...
+QuizOptions.TRUE_OR_FALSE = "True or False";
+QuizOptions.QUESTION_TYPES =
+    [  QuizOptions.MULTIPLE_CHOICE_SINGLE_ANSWER,
+        QuizOptions.MULTIPLE_CHOICE_MULTIPLE_ANSWERS,
+        QuizOptions.TRUE_OR_FALSE
+    ];
+
+
+
+QuizQuestion = function() {};
```
Comment (brylie) [inline both/quizModel.js:114]: SimpleSchema provides the structure for Questions, while JavaScript provides simple access to the object properties without the need for getters or setters.

File: both/quizModel.js [hunk 30]
```diff
@@ -0,0 +1,197 @@
...
+    get: function questionType() {
+        return this._quiztype;
+    },
+    set: function questionType(val) {
+        this._quiztype = val;
+    }
+});
+
+//array of lesson IDs - the quiz can be used in more than one lesson
+Object.defineProperty(QuizQuestion, "_quizId", {
```
Comment (brylie) [inline both/quizModel.js:125]: This is handled by SimpleSchema.

File: both/quizModel.js [hunk 31]
```diff
@@ -0,0 +1,197 @@
...
+    get: function quizId(){
+        return this._quizId
+    },
+    set: function quizId(val){
+        this._lessonIDs = val;
+    }
+});
+
+Object.defineProperty(QuizQuestion, "_title", {
```
Comment (brylie) [inline both/quizModel.js:134]: This is handled by SimpleSchema. JavaScript properties are accessible on returned MongoDB document objects.

File: both/quizModel.js [hunk 32]
```diff
@@ -0,0 +1,197 @@
...
+    get: function title(){
+        return this._title;
+    },
+    set: function title(val){
+        this._title = val;
+    }
+});
+
+Object.defineProperty(QuizQuestion, "_description", {
```
Comment (brylie) [inline both/quizModel.js:143]: This is handled by SimpleSchema. JavaScript properties are accessible on returned MongoDB document objects.

File: both/quizModel.js [hunk 33]
```diff
@@ -0,0 +1,197 @@
...
+    get: function description(){
+        return this._description;
+    },
+    set: function description(val){
+        this._description = val;
+    }
+});
+
+Object.defineProperty(QuizQuestion, "_options", {
```
Comment (brylie) [inline both/quizModel.js:152]: This is handled by SimpleSchema. JavaScript properties are accessible on returned MongoDB document objects.

File: both/quizModel.js [hunk 34]
```diff
@@ -0,0 +1,197 @@
...
+    get: function options(){
+        return this._options;
+    },
+    set: function options(val){
+        this._questions = val;
+    }
+});
+
+Object.defineProperty(QuizQuestion, "_saved",{
```
Comment (brylie) [inline both/quizModel.js:161]: This is handled by SimpleSchema. JavaScript properties are accessible on returned MongoDB document objects.

File: both/quizModel.js [hunk 35]
```diff
@@ -0,0 +1,197 @@
...
+    get: function saved(){
+        return this._saved;
+    },
+    set: function saved(val){
+        this._saved = val;
+    }
+});
+
+Object.defineProperty(QuizQuestion, "_answered", {
```
Comment (brylie) [inline both/quizModel.js:170]: This is handled by SimpleSchema. JavaScript properties are accessible on returned MongoDB document objects.

File: both/quizModel.js [hunk 36]
```diff
@@ -0,0 +1,197 @@
...
+    get: function answered(){
+        return this._answered;
+    },
+    set: function answered(val){
+        this._answered = val;
+    }
+})
+
+Object.defineProperty(QuizQuestion, "_isMultipleAnswer", {
```
Comment (brylie) [inline both/quizModel.js:179]: This is handled by SimpleSchema. JavaScript properties are accessible on returned MongoDB document objects.

File: both/quizModel.js (hunk 30 above)
Comment (zeala) [inline both/quizModel.js:125]: The quizModel class was left in the code by mistake (it was used in the previous implementation - prior to the SimpleSchema and autoform integration).

I'll remove the file.

File: client/helpers/editMode.js [hunk 37]
```diff
@@ -2,3 +2,20 @@ Template.registerHelper('editMode', function () {
     // get edit mode session variable
     return Session.get('editMode');
 });
+
+Template.registerHelper('isEditingCurrentCourse', function() {
```
Comment (brylie) [inline client/helpers/editMode.js:6]: Good idea to move this to a general Template helper, since it was duplicated.

File: client/templates/course/lesson/lesson.css [hunk 38]
```diff
@@ -0,0 +1,56 @@
+
+.add-quiz-text {
+  cursor: pointer;
+  font-size: 11px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.css:4]: Use relative values such as `em` or `rem`.

File: client/templates/course/lesson/lesson.css [hunk 39]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+.add-quiz-btn {
+    color: #FFFFFF;
+    background-color: #5CB85C;
+    border: #bbb 4px solid;
+    border-radius: 4px;
+    margin: 5px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.css:12]: For layout related styling, use relative values such as `em` or `rem`.

File: client/templates/course/lesson/lesson.css [hunk 40]
```diff
@@ -0,0 +1,56 @@
...
+    margin-right: 10px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.css:13]: For layout related styling, use relative values such as `em` or `rem`.

File: client/templates/course/lesson/lesson.css [hunk 41]
```diff
@@ -0,0 +1,56 @@
...
+    width: 100%;
+    padding: 3px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.css:15]: For layout related styling, use relative values such as `em` or `rem`.

File: client/templates/course/lesson/lesson.html [hunk 42]
```diff
@@ -1,14 +1,20 @@
...
+                </div>
             </div>
         </div>
-    </div>
-    {{/ with }}
+        {{/ with }}
+    {{/ if }}
+
+    {{# if activeQuiz }}
+        {{> quizContent activeQuiz=activeQuiz }}
```
Comment (brylie) [inline client/templates/course/lesson/lesson.html:18]: Good idea to explicitly pass the `activeQuiz` context here. :-)

File: client/templates/course/lesson/lesson.js [hunk 43]
```diff
@@ -7,5 +7,12 @@ Template.lesson.helpers({
     var lesson = Lessons.findOne({_id: lessonID});
 
     return lesson;
+  },
+
+  'activeQuiz': function(){
+    var quizId = activeQuizID.get();
```
Comment (brylie) [inline client/templates/course/lesson/lesson.js:13]: Where is `activeQuizID` defined? Is it a `Template.instance` variable?

File: client/templates/course/lesson/lesson.js [hunk 44]
```diff
@@ -7,5 +7,12 @@ Template.lesson.helpers({
...
+
+    var quiz = Quiz.convertToQuizObject(Quizzes.findOne({ _id: quizId }));
```
Comment (brylie) [inline client/templates/course/lesson/lesson.js:15]: This may be unnecessary, as results of `Collection.findOne()` are already objects.

File: client/templates/course/lesson/lesson.less [hunk 45]
```diff
@@ -0,0 +1,9 @@
+#addQuizBtn{
+  margin-left: 12px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.less:2]: For layout related styling, use relative values such as `em` or `rem`.

File: client/templates/course/lesson/lesson.less [hunk 46]
```diff
@@ -0,0 +1,9 @@
...
+  margin-top: 5px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.less:3]: For layout related styling, use relative values such as `em` or `rem`.

File: client/templates/course/lesson/lesson.less [hunk 47]
```diff
@@ -0,0 +1,9 @@
...
+}
+
+.add-quiz-text{
+  cursor: pointer;
+  font-size: 11px;
```
Comment (brylie) [inline client/templates/course/lesson/lesson.less:8]: For layout related styling, use relative values such as `em` or `rem`.

File: client/templates/course/lesson/quiz/questions/autoform.js [hunk 48]
```diff
@@ -0,0 +1,26 @@
+
+var addQuestionFormHooks = {
+    before: {
+        method: function(doc, params){
+            var editedQuestion = Session.get("currentQuestionToBuild");
+            editedQuestion.title = doc.title;
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform.js:6]: AutoForm already handles this document properties, please stay closer to AutoForm when managing document properties.

File: client/templates/course/lesson/quiz/questions/autoform.js [hunk 49]
```diff
@@ -0,0 +1,26 @@
+
+var addQuestionFormHooks = {
+    before: {
+        method: function(doc, params){
+            var editedQuestion = Session.get("currentQuestionToBuild");
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform.js:5]: Use session variables sparingly, and generally only to hold strings.

File: client/templates/course/lesson/quiz/questions/autoform.js [hunk 50]
```diff
@@ -0,0 +1,26 @@
...
+            editedQuestion.title = doc.title;
+            editedQuestion.description = doc.description;
+            return editedQuestion;
+        }
+    },
+    onSuccess: function(operation, result, template) {
+
+        //clear the session cache of the currently edited question
+        Session.set("currentQuestionToBuild", undefined);
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform.js:14]: All of these session variables may prove unnecessary, if you embrace AutoForm for this feature.

File: client/templates/course/lesson/quiz/questions/autoform.js [hunk 51]
```diff
@@ -0,0 +1,26 @@
...
+        Session.set("currentQuestionTypeChanged", undefined);
+        delete Session.keys["currentQuestionToBuild"];
+        delete Session.keys["currentQuestionTypeChanged"];
+        return sAlert.success("question added");
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform.js:18]: The `return` statement may be unnecessary here. Simply using `sAlert...` may suffice.

File: client/templates/course/lesson/quiz/questions/autoform/afQuestionOptionsGroup.html [hunk 52]
```diff
@@ -0,0 +1,26 @@
+<template name="afFieldQuestionOptionsGroup">
+    {{#each displayedOptions }}
+        <div style="{{#if isHidden}}display: none{{/if}}" >
+            <span class="radio-inline checkbox-inline">
+                    {{#if isMultipleAnswer }}
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform/afQuestionOptionsGroup.html:5]: This is a good approach, and can be generalized to show specific fields based on question type:

``` js
{{# if afFieldValueIs name="question.type" value="multipleChoice" }}
  <!-- Show the multiple choice question field -->
  {{> afQuickField name="multipleChoice" }}
{{/ if }}
```

Then, each of the field types would be defined with their own sub-schema. This means we can easily add new question types in the future.

File: client/templates/course/lesson/quiz/questions/autoform/afQuestionOptionsGroup.js [hunk 53]
```diff
@@ -0,0 +1,123 @@
+Template.afFieldQuestionOptionsGroup.helpers({
+
+    displayedOptions: function(){
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform/afQuestionOptionsGroup.js:3]: These helpers may be unnecessarily complicated. E.g. simply querying for question options, and omitting the `isCorrectAnswer` field would provide an array of question options to iterate through. Hence, the transformations would be unnecessary within the helper.

File: client/templates/course/lesson/quiz/questions/autoform/afQuestionOptionsGroup.js [hunk 54]
```diff
@@ -0,0 +1,123 @@
...
+        var totalQuestions = question && question.optionTitles ? question.optionTitles.length : 0;
+
+        for (var i= 0; i< totalQuestions; i++){
+            displayedOptions.push (allOptions[i]);
+        }
+
+        return displayedOptions;
+    },
+
+    isHidden: function(){
```
Comment (brylie) [inline client/templates/course/lesson/quiz/questions/autoform/afQuestionOptionsGroup.js:18]: Again, overly complicated. The `isHidden` question property may not prove useful in the long term.

**Comment by brylie** [comment]
Lets break this feature effort into multiple parts:
- Defining collection, schema, and AutoForm for inputting the quiz data
- Defining templates and helpers for displaying quizzes to end users and tracking their submissions
- Defining logic to compare user submissions with correct answers

The code in this branch is really complicated, and could be greatly simplified by focusing on the collection, schema, and autoform for quiz definition.

**Comment by brylie** [comment]
Also, please refer to the initial schema suggestion (with some modification):

``` js
QuizQuestions = new Mongo.Collection("quizQuestions");
QuizQuestions.schema = new SimpleSchema({
    questionType: {
       type: String,
        allowedValues: ['multipleChoice', 'freeform', ...]
        // Extend this list as new question types are added
    },
    multiChoice: {
      type: Object
      optional: true,
      customValidation: function () {
         // field is required only if quizType is set to 'multipleChoice'
         // Get value of questionType field
         let questionType = this.field("questionType").value
        if (questionType === "multipleChoice") {
            return true; // field is required
        } else {
          return false; // field is not required
        }
       }
    },
    "multiChoice.question" { // Question text to display to the user
        type: String
    }
    "multiChoice.choice": {
        type: [Object], // brackets here mean the question can have multiple choice
    },
    "multiChoice.choice.text": {
        type: String // Text to show user, i.e. answer text
    },
    "multiChoice.choice.isCorrectAnswer": {
        type: Boolean // It may be possible to have multiple correct answers
    }
});
```

With this schema, we can define new question types easily. We can then use AutoForm `afFieldValueIs` to dynamically toggle the sub fields based on the `quizType` field value on the quiz input form:

``` js
<!-- Create new quiz form, add question section -->
{{# if afFieldValueIs name="questionType" value="multipleChoice" }}
  <!-- Show the multiple choice question field -->
  {{> afQuickField name="multipleChoice" }}
{{/ if }}
```

**Comment by brylie** [comment]
I think we can shrink this PR to closer to 500 lines of code, with re-design. The current state of the code can be preserved in this branch, if you simply create a new feature branch. E.g. `feature/quiz-input-form-collection-and-schema`.
